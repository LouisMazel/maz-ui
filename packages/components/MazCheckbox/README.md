# MazCheckbox

## Props

<!-- @vuese:MazCheckbox:props:start -->

| Name  | Description | Type      | Required | Default      |
| ----- | ----------- | --------- | -------- | ------------ |
| value | -           | `Boolean` | `true`   | -            |
| id    | -           | `String`  | `false`  | null         |
| color | -           | `String`  | `false`  | primary      |
| name  | -           | `String`  | `false`  | maz-checkbox |

<!-- @vuese:MazCheckbox:props:end -->

## Events

<!-- @vuese:MazCheckbox:events:start -->

| Event Name | Description | Parameters |
| ---------- | ----------- | ---------- |
| input      | -           | -          |

<!-- @vuese:MazCheckbox:events:end -->

## Slots

<!-- @vuese:MazCheckbox:slots:start -->

| Name    | Description | Default Slot Content |
| ------- | ----------- | -------------------- |
| default | -           | -                    |

<!-- @vuese:MazCheckbox:slots:end -->

## MixIns

<!-- @vuese:MazCheckbox:mixIns:start -->

| MixIn    |
| -------- |
| uniqueId |

<!-- @vuese:MazCheckbox:mixIns:end -->
