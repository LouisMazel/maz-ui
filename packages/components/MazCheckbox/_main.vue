<template>
  <div
    class="maz-base-component maz-checkbox"
    :class="[`maz-checkbox--${color}`]"
  >
    <input
      :id="uniqueId"
      :checked="value"
      v-bind="$attrs"
      :name="name"
      type="checkbox"
      class="maz-mr-2"
      @change="$emit('input', $event.target.checked)"
    >
    <label
      :for="uniqueId"
      class="maz-m-0 maz-flex maz-align-center"
    >
      <slot />
    </label>
  </div>
</template>

<script>
import uniqueId from './../../mixins/uniqueId'

export default {
  name: 'MazCheckbox',
  mixins: [uniqueId],
  props: {
    value: { type: Boolean, required: true },
    id: { type: String, default: null },
    color: { type: String, default: 'primary' },
    name: { type: String, default: 'maz-checkbox'}
  }
}
</script>
