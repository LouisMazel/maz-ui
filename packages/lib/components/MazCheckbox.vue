<template>
  <label
    :for="instanceId"
    class="m-checkbox"
    :class="[{ '--disabled': disabled }, props.class]"
    tabindex="0"
    :style="style"
    role="checkbox"
    :aria-checked="isChecked"
    @keydown="keyboardHandler"
  >
    <input
      :id="instanceId"
      :checked="isChecked"
      v-bind="$attrs"
      tabindex="-1"
      :disabled="disabled"
      :name="name"
      type="checkbox"
      @change="emitValue(value ?? ($event?.target as HTMLInputElement)?.checked)"
    />
    <span>
      <CheckIcon class="check-icon" :class="checkIconSize" />
    </span>
    <slot></slot>
  </label>
</template>

<script lang="ts" setup>
  import { useInstanceUniqId } from '../modules/composables/use-instance-uniq-id'
  import { getCurrentInstance, computed, type HTMLAttributes } from 'vue'
  import type { Color, Size } from './types'
  export type { Color, Size }
  import CheckIcon from './../icons/check.svg'

  const instance = getCurrentInstance()

  defineOptions({
    inheritAttrs: false,
  })

  export type Props = {
    style?: HTMLAttributes['style']
    class?: HTMLAttributes['class']
    modelValue?: boolean | (string | number)[]
    id?: string
    color?: Color
    value?: string | number | boolean
    name?: string
    size?: Size
    disabled?: boolean
  }

  const props = withDefaults(defineProps<Props>(), {
    style: undefined,
    class: undefined,
    modelValue: undefined,
    id: undefined,
    color: 'primary',
    value: undefined,
    name: 'm-checkbox',
    size: 'md',
    disabled: false,
  })

  const emits = defineEmits([
    /* emitted when value change */
    'update:model-value',
    /* emited when value change */
    'change',
  ])

  const instanceId = useInstanceUniqId({
    componentName: 'MazCheckbox',
    instance,
    providedId: props.id,
  })

  const isChecked = computed(() =>
    typeof props.value !== 'boolean' && Array.isArray(props.modelValue)
      ? props.modelValue.includes(props.value as never)
      : typeof props.modelValue === 'boolean'
        ? props.modelValue
        : false,
  )

  const checkboxSize = computed(() => {
    switch (props.size) {
      case 'xl': {
        return '2.25rem'
      }
      case 'lg': {
        return '2rem'
      }
      default: {
        return '1.625rem'
      }
      case 'sm': {
        return '1.425rem'
      }
      case 'xs': {
        return '1.325rem'
      }
      case 'mini': {
        return '1.2rem'
      }
    }
  })

  const checkIconSize = computed(() => {
    switch (props.size) {
      case 'xl': {
        return 'maz-text-2xl'
      }
      case 'lg': {
        return 'maz-text-xl'
      }
      default: {
        return 'maz-text-lg'
      }
      case 'sm': {
        return 'maz-text-base'
      }
      case 'xs': {
        return 'maz-text-sm'
      }
      case 'mini': {
        return 'maz-text-xs'
      }
    }
  })

  const checkIconColor = computed(() => `var(--maz-color-${props.color}-contrast)`)
  const checkboxSelectedColor = computed(() => `var(--maz-color-${props.color})`)
  const checkboxBoxShadow = computed(() =>
    ['black', 'transparent'].includes(props.color)
      ? `var(--maz-color-muted)`
      : `var(--maz-color-${props.color}-alpha)`,
  )

  function keyboardHandler(event: KeyboardEvent) {
    if (['Space'].includes(event.code)) {
      event.preventDefault()
      emitValue(props.value ?? !props.modelValue)
    }
  }

  function getNewValue(value: boolean | string | number) {
    if (
      typeof value === 'boolean' &&
      (typeof props.modelValue === 'boolean' ||
        props.modelValue === undefined ||
        props.modelValue === null)
    ) {
      return props.modelValue ? false : true
    } else if (Array.isArray(props.modelValue) && typeof value !== 'boolean') {
      return props.modelValue.includes(value)
        ? props.modelValue.filter((v) => v !== value)
        : [...props.modelValue, value]
    } else {
      return [value]
    }
  }

  function emitValue(value: boolean | string | number) {
    const newValue = getNewValue(value)

    emits('update:model-value', newValue)
    emits('change', newValue)
  }
</script>

<style lang="postcss" scoped>
  .m-checkbox {
    @apply maz-relative maz-inline-flex maz-items-center maz-gap-2 maz-outline-none;

    .check-icon {
      color: v-bind('checkIconColor');

      @apply maz-scale-0 maz-transition-transform maz-duration-300 maz-ease-in-out;

      :deep(path) {
        stroke-width: 2.5;
      }
    }

    > span {
      @apply maz-relative maz-flex maz-rounded-md maz-border maz-border-border maz-transition-all maz-duration-300 maz-ease-in-out maz-flex-center dark:maz-border-color-lighter;

      width: v-bind('checkboxSize');
      height: v-bind('checkboxSize');
    }

    input {
      @apply maz-hidden;

      &:not(:checked) ~ span {
        @apply maz-bg-color dark:maz-bg-color-light;
      }

      &:checked ~ span {
        border-color: v-bind('checkboxSelectedColor');
        background-color: v-bind('checkboxSelectedColor');

        .check-icon {
          @apply maz-scale-100;
        }
      }

      &:disabled ~ span {
        @apply maz-bg-color-light dark:maz-bg-color-lighter;
      }
    }

    &.--disabled {
      @apply maz-cursor-not-allowed maz-text-muted;

      input:checked ~ span {
        @apply maz-border-border dark:maz-border-color-lighter;

        .check-icon {
          @apply maz-text-muted;
        }
      }
    }

    &:not(.--disabled) {
      @apply maz-cursor-pointer;

      &:hover > span,
      &:focus > span {
        @apply maz-transition-all maz-duration-300 maz-ease-in-out;

        box-shadow: 0 0 0 0.125rem v-bind('checkboxBoxShadow');
      }
    }
  }
</style>
