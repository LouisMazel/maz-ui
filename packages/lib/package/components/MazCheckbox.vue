<template>
  <div class="m-checkbox" :class="[`m-checkbox--${color}`]">
    <input
      :id="id"
      :checked="modelValue"
      v-bind="$attrs"
      :name="name"
      type="checkbox"
      class="maz-mr-2"
      @change="$emit('update:model-value', ($event?.target as HTMLInputElement)?.checked)"
    />
    <label :for="id" class="maz-m-0 maz-flex maz-items-center">
      <slot />
    </label>
  </div>
</template>

<script lang="ts">
  export type { Color } from './types'
</script>

<script lang="ts" setup>
  import { PropType } from 'vue'
  import type { Color } from './types'

  defineProps({
    modelValue: { type: Boolean, required: true },
    id: { type: String, default: 'MazCheckbox' },
    color: {
      type: String as PropType<Color>,
      default: 'primary',
      validator: (value: Color) => {
        return [
          'primary',
          'secondary',
          'info',
          'success',
          'warning',
          'danger',
          'white',
          'black',
          'transparent',
        ].includes(value)
      },
    },
    name: { type: String, default: 'm-checkbox' },
  })

  defineEmits(['update:model-value'])
</script>

<style lang="postcss" scoped>
  .m-checkbox {
    transition: all 300ms ease-in-out;
    cursor: pointer;
    margin-left: 2px;
    min-height: 22px;

    [type='checkbox']:not(:checked),
    [type='checkbox']:checked {
      position: absolute;
      left: -9999px;
    }

    [type='checkbox']:not(:checked) + label,
    [type='checkbox']:checked + label {
      position: relative;
      padding-left: 30px;
      font-size: 1rem;
      cursor: pointer;
      transition: all 300ms ease-in-out;
      user-select: none;
    }

    [type='checkbox'] + label::before {
      @apply maz-border maz-border-solid maz-border-transparent;

      content: '';
      position: absolute;
      left: 0;
      width: 18px;
      height: 18px;
      background: transparent;
      border-radius: 4px;
      transition: all 300ms ease-in-out;
    }

    [type='checkbox']:focus + label::before {
      @apply maz-rounded maz-border maz-border-solid maz-border-primary;

      content: '';
      position: absolute;
      left: 0;
      width: 18px;
      height: 18px;
      background: transparent;
      transition: all 300ms ease-in-out;
    }

    [type='checkbox']:not(:checked) + label::before {
      @apply maz-border-color-light;
    }

    [type='checkbox']:not(:checked) + label::after,
    [type='checkbox']:checked + label::after {
      content: '';
      position: absolute;
      left: 5px;
      width: 12px;
      height: 12px;
      transition: all 300ms ease-in-out;
      border-radius: 2px;
    }

    [type='checkbox']:not(:checked) + label::after {
      opacity: 0;
      transform: scale(0);
    }

    [type='checkbox']:checked + label::after {
      opacity: 1;
      transform: scale(1);
    }

    &--primary {
      [type='checkbox']:checked + label::before {
        @apply maz-border-primary;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-primary maz-text-primary;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-primary-alpha);

          @apply maz-border-primary-700;
        }

        + label::after {
          @apply maz-bg-primary-700;
        }
      }
    }

    &--secondary {
      [type='checkbox']:checked + label::before {
        @apply maz-border-secondary;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-secondary maz-text-secondary;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-secondary-alpha);

          @apply maz-border-secondary-700;
        }

        + label::after {
          @apply maz-bg-secondary-700;
        }
      }
    }

    &--info {
      [type='checkbox']:checked + label::before {
        @apply maz-border-info;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-info maz-text-info;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-info-alpha);

          @apply maz-border-info-700;
        }

        + label::after {
          @apply maz-bg-info-700;
        }
      }
    }

    &--danger {
      [type='checkbox']:checked + label::before {
        @apply maz-border-danger;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-danger maz-text-danger;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-danger-alpha);

          @apply maz-border-danger-700;
        }

        + label::after {
          @apply maz-bg-danger-700;
        }
      }
    }

    &--warning {
      [type='checkbox']:checked + label::before {
        @apply maz-border-warning;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-warning maz-text-warning;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-warning-alpha);

          @apply maz-border-warning-700;
        }

        + label::after {
          @apply maz-bg-warning-700;
        }
      }
    }

    &--success {
      [type='checkbox']:checked + label::before {
        @apply maz-border-success;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-success maz-text-success;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem var(--maz-color-success-alpha);

          @apply maz-border-success-700;
        }

        + label::after {
          @apply maz-bg-success-700;
        }
      }
    }

    &--white {
      [type='checkbox']:checked + label::before {
        @apply maz-border-white;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-white maz-text-white;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem rgb(209 213 219 / 60%);

          @apply maz-border-gray-300;
        }

        + label::after {
          @apply maz-bg-gray-300;
        }
      }
    }

    &--black {
      [type='checkbox']:checked + label::before {
        @apply maz-border-black;
      }

      [type='checkbox']:not(:checked) + label::after,
      [type='checkbox']:checked + label::after {
        @apply maz-bg-black maz-text-black;
      }

      [type='checkbox']:focus {
        + label::before {
          box-shadow: 0 0 0 0.143rem rgb(0 0 0 / 60%);

          @apply maz-border-gray-700;
        }

        + label::after {
          @apply maz-bg-gray-700;
        }
      }
    }
  }

  html.dark {
    .m-checkbox {
      [type='checkbox']:not(:checked) + label::before {
        @apply maz-border-color-lighter;
      }
    }
  }
</style>
