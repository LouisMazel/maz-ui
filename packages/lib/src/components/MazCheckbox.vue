<script lang="ts" setup generic="T extends boolean | (string | number)[]">
import type { Color, Size } from './types'
import { computed, type HTMLAttributes, ref } from 'vue'
import CheckIcon from '../../icons/check.svg'
import { useInstanceUniqId } from '../composables/useInstanceUniqId'

export type CheckboxValue = string | number | boolean

export interface MazCheckboxProps<T = boolean | (string | number)[]> {
  /** Style attribut of the component root element */
  style?: HTMLAttributes['style']
  /** Class attribut of the component root element */
  class?: HTMLAttributes['class']
  /** The model value of the checkbox */
  modelValue?: T
  /** The id of the checkbox */
  id?: string
  /** The color of the checkbox */
  color?: Color
  /** The value of the checkbox when selected */
  value?: CheckboxValue
  /** The name of the checkbox */
  name?: string
  /** The size of the checkbox */
  size?: Size
  /** Text label */
  label?: string
  /** If the checkbox is disabled */
  disabled?: boolean
  /** Whether there is an error with the input. */
  error?: boolean
  /** Whether the input is successful. */
  success?: boolean
  /** Whether there is a warning with the input. */
  warning?: boolean
  /** The hint text to display below the input. */
  hint?: string
}

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(
  defineProps<MazCheckboxProps<T>>(),
  {
    style: undefined,
    class: undefined,
    modelValue: undefined,
    label: undefined,
    id: undefined,
    color: 'primary',
    value: undefined,
    name: 'm-checkbox',
    size: 'md',
    disabled: false,
  },
)

const emits = defineEmits<{
  /**
   * Emitted when the model value change
   * @property value The new value
   */
  'update:model-value': [value: T]
  /**
   * Emitted when the model value change
   * @property value The new value
   */
  'change': [value: T]
  /**
   * Emitted when the checkbox lost focus
   * @property {FocusEvent} value - The focus event
   */
  'blur': [value: FocusEvent]
  /**
   * Emitted when the checkbox is focused
   * @property {FocusEvent} value - The focus event
   */
  'focus': [value: FocusEvent]
}>()

const instanceId = useInstanceUniqId({
  componentName: 'MazCheckbox',
  providedId: props.id,
})

const isChecked = computed(() =>
  typeof props.value !== 'boolean' && Array.isArray(props.modelValue)
    ? props.modelValue.includes(props.value as never)
    : typeof props.modelValue === 'boolean'
      ? props.modelValue
      : false,
)

const checkboxSize = computed(() => {
  switch (props.size) {
    case 'xl': {
      return '2rem'
    }
    case 'lg': {
      return '1.75rem'
    }
    case 'sm': {
      return '1.25rem'
    }
    case 'xs': {
      return '1rem'
    }
    case 'mini': {
      return '0.75rem'
    }

    default: {
      return '1.5rem'
    }
  }
})

const checkIconSize = computed(() => {
  switch (props.size) {
    case 'xl': {
      return 'maz-text-2xl'
    }
    case 'lg': {
      return 'maz-text-xl'
    }
    case 'sm': {
      return 'maz-text-base'
    }
    case 'xs': {
      return 'maz-text-sm'
    }
    case 'mini': {
      return 'maz-text-xs'
    }
    default: {
      return 'maz-text-lg'
    }
  }
})

const checkIconColor = computed(() => {
  if (props.color === 'theme') {
    return 'var(--maz-color-bg)'
  }

  return `var(--maz-color-${props.color}-contrast)`
})
const checkboxSelectedColor = computed(() => {
  if (props.color === 'theme') {
    return 'var(--maz-color-bg-theme)'
  }

  return `var(--maz-color-${props.color})`
})
const checkboxBoxShadow = computed(() => {
  if (props.error) {
    return `var(--maz-color-danger)`
  }
  else if (props.warning) {
    return `var(--maz-color-warning)`
  }
  else if (props.success) {
    return `var(--maz-color-success)`
  }

  return ['black', 'transparent', 'theme'].includes(props.color)
    ? `var(--maz-color-muted)`
    : `var(--maz-color-${props.color}-alpha)`
})

function keyboardHandler(event: KeyboardEvent) {
  if (['Space'].includes(event.code)) {
    event.preventDefault()
    emitValue(props.value ?? !props.modelValue)
  }
}

function getNewValue(value: boolean | string | number) {
  if (
    typeof value === 'boolean'
    && (typeof props.modelValue === 'boolean'
      || props.modelValue === undefined
      || props.modelValue === null)
  ) {
    return !props.modelValue
  }
  else if (Array.isArray(props.modelValue) && typeof value !== 'boolean') {
    return props.modelValue.includes(value)
      ? props.modelValue.filter(v => v !== value)
      : [...props.modelValue, value]
  }
  else {
    return [value]
  }
}

function emitValue(value: boolean | string | number) {
  const newValue = getNewValue(value)

  emits('update:model-value', newValue as T)
  emits('change', newValue as T)
}

const inputRef = ref<HTMLInputElement>()

function onBlur(event: FocusEvent) {
  inputRef.value?.dispatchEvent(new Event('blur'))
  emits('blur', event)
}
function onFocus(event: FocusEvent) {
  inputRef.value?.dispatchEvent(new Event('focus'))
  emits('focus', event)
}
</script>

<template>
  <label
    :for="instanceId"
    class="m-checkbox m-reset-css"
    :class="[{ '--disabled': disabled, '--error': error, '--warning': warning, '--success': success }, props.class]"
    tabindex="0"
    :style="[style, { '--checkbox-selected-color': checkboxSelectedColor, '--checkbox-box-shadow-color': checkboxBoxShadow }]"
    role="checkbox"
    :aria-checked="isChecked"
    @keydown="keyboardHandler"
    @blur="onBlur"
    @focus="onFocus"
  >
    <input
      :id="instanceId"
      ref="inputRef"
      :checked="isChecked"
      v-bind="$attrs"
      tabindex="-1"
      :disabled
      :name
      type="checkbox"
      @change="emitValue(value ?? ($event?.target as HTMLInputElement)?.checked)"
    >
    <span :style="{ width: checkboxSize, height: checkboxSize }">
      <CheckIcon class="check-icon" :class="checkIconSize" :style="{ color: checkIconColor }" />
    </span>
    <div class="m-checkbox__text">
      <slot :value>
        {{ label }}
      </slot>

      <span
        v-if="hint"
        class="m-checkbox__hint" :class="{
          '--error': error,
          '--success': success,
          '--warning': warning,
        }"
        :style="{ boxShadow: `0 0 0 0.125rem ${checkboxBoxShadow}` }"
      >{{ hint }}</span>
    </div>
  </label>
</template>

<style lang="postcss" scoped>
  .m-checkbox {
  @apply maz-relative maz-inline-flex maz-items-center maz-gap-2 maz-align-top maz-outline-none;

  .check-icon {
    @apply maz-scale-0 maz-transition-transform maz-duration-300 maz-ease-in-out;

    :deep(path) {
      stroke-width: 2.5;
    }
  }

  > span {
    @apply maz-relative maz-flex maz-rounded-md maz-border maz-border-border maz-transition-all maz-duration-300 maz-ease-in-out maz-flex-center dark:maz-border-color-lighter;
  }

  input {
    @apply maz-hidden;

    &:not(:checked) ~ span {
      @apply maz-bg-color dark:maz-bg-color-light;
    }

    &:checked ~ span {
      border-color: var(--checkbox-selected-color);
      background-color: var(--checkbox-selected-color);

      .check-icon {
        @apply maz-scale-100;
      }
    }

    &:disabled ~ span {
      @apply maz-bg-color-light dark:maz-bg-color-lighter;
    }
  }

  &.--disabled {
    @apply maz-cursor-not-allowed maz-text-muted;

    input:checked ~ span {
      @apply maz-border-border dark:maz-border-color-lighter;

      .check-icon {
        @apply maz-text-muted;
      }
    }
  }

  &:not(.--disabled) {
    @apply maz-cursor-pointer;

    &:hover > span,
    &:focus > span {
      @apply maz-transition-all maz-duration-300 maz-ease-in-out;

      box-shadow: 0 0 0 0.125rem var(--checkbox-box-shadow-color);
    }
  }

  &__text {
    @apply maz-flex maz-flex-col maz-gap-0;
  }

  &__hint {
    @apply maz-text-sm maz-text-muted;

    &.--error {
      @apply maz-text-danger-600;
    }

    &.--success {
      @apply maz-text-success-600;
    }

    &.--warning {
      @apply maz-text-warning-600;
    }
  }

  &.--error,
  &.--warning,
  &.--success {
    > span {
      @apply maz-transition-all maz-duration-300 maz-ease-in-out;
    }
  }
}
</style>
