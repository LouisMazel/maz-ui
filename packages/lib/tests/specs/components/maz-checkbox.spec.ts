import MazCheckbox from '@components/MazCheckbox.vue'
import { mount } from '@vue/test-utils'

describe('components/MazCheckbox.vue', () => {
  expect(MazCheckbox).toBeTruthy()

  it('renders correctly', () => {
    const wrapper = mount(MazCheckbox, {
      props: {
        modelValue: true, // You can provide props here
        // Add other props as needed
      },
    })

    // Add your assertions based on your component's rendering and behavior
    // For example, you can assert the existence of specific elements or classes:
    expect(wrapper.find('label').exists()).toBe(true)
    expect(wrapper.find('input').attributes('type')).toBe('checkbox')
    // Add more assertions as needed.

    // Don't forget to clean up the wrapper
    wrapper.unmount()
  })

  it('should have an uniq id', () => {
    const wrapper = mount(MazCheckbox, {
      props: {
        modelValue: false,
      },
    })

    expect(wrapper.vm.instanceId).toBe('MazCheckbox-4')

    wrapper.unmount()
  })
})
