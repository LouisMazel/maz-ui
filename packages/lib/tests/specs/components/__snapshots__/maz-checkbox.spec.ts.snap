// Vitest Snapshot v1

exports[`components/MazCheckbox.vue > Should match with the snapshot 1`] = `"<div class=\\"m-checkbox m-checkbox--primary\\" data-v-040cc227=\\"\\"><input id=\\"MazCheckbox-1\\" name=\\"m-checkbox\\" type=\\"checkbox\\" class=\\"maz-mr-2\\" data-v-040cc227=\\"\\"><label for=\\"MazCheckbox-1\\" class=\\"maz-m-0 maz-flex maz-items-center\\" data-v-040cc227=\\"\\"></label></div>"`;

exports[`components/MazCheckbox.vue 1`] = `"<div class=\\"m-checkbox m-checkbox--primary\\" data-v-040cc227=\\"\\"><input id=\\"MazCheckbox-1\\" name=\\"m-checkbox\\" type=\\"checkbox\\" class=\\"maz-mr-2\\" data-v-040cc227=\\"\\"><label for=\\"MazCheckbox-1\\" class=\\"maz-m-0 maz-flex maz-items-center\\" data-v-040cc227=\\"\\"></label></div>"`;
